File: cctv.js
// ==================== 央视大全爬虫 (简化版，直接使用原始 hls_url) ====================
let globalHeaders = {
  "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.54 Safari/537.36",
  "Origin": "https://tv.cctv.com",
  "Referer": "https://tv.cctv.com/"
};

let cache = {};

function fetchSync(url, options = {}) {
  if (cache[url]) return cache[url];
  try {
    let reqOpts = { method: options.method || 'GET', headers: { ...globalHeaders, ...(options.headers || {}) } };
    if (options.data && reqOpts.method === 'POST') {
      reqOpts.body = JSON.stringify(options.data);
      reqOpts.headers['Content-Type'] = 'application/json';
    }
    let resp = req(url, reqOpts);
    let content = typeof resp === 'string' ? resp : resp.content;
    if (options.json) {
      let json = JSON.parse(content);
      if (options.cache !== false) cache[url] = json;
      return json;
    }
    if (options.cache !== false) cache[url] = content;
    return content;
  } catch (e) {
    console.log(`请求失败: ${url} - ${e.message}`);
    return options.json ? null : '';
  }
}

// 筛选器配置（完整保留）
const filtersConfig = [
  {
    "key": "cid", "name": "频道", "value": [
      { "n": "全部", "v": "" },
      { "n": "CCTV-1综合", "v": "EPGC1386744804340101" },
      { "n": "CCTV-2财经", "v": "EPGC1386744804340102" },
      { "n": "CCTV-3综艺", "v": "EPGC1386744804340103" },
      { "n": "CCTV-4中文国际", "v": "EPGC1386744804340104" },
      { "n": "CCTV-5体育", "v": "EPGC1386744804340107" },
      { "n": "CCTV-6电影", "v": "EPGC1386744804340108" },
      { "n": "CCTV-7国防军事", "v": "EPGC1386744804340109" },
      { "n": "CCTV-8电视剧", "v": "EPGC1386744804340110" },
      { "n": "CCTV-9纪录", "v": "EPGC1386744804340112" },
      { "n": "CCTV-10科教", "v": "EPGC1386744804340113" },
      { "n": "CCTV-11戏曲", "v": "EPGC1386744804340114" },
      { "n": "CCTV-12社会与法", "v": "EPGC1386744804340115" },
      { "n": "CCTV-13新闻", "v": "EPGC1386744804340116" },
      { "n": "CCTV-14少儿", "v": "EPGC1386744804340117" },
      { "n": "CCTV-15音乐", "v": "EPGC1386744804340118" },
      { "n": "CCTV-16奥林匹克", "v": "EPGC1634630207058998" },
      { "n": "CCTV-17农业农村", "v": "EPGC1563932742616872" },
      { "n": "CCTV-5+体育赛事", "v": "EPGC1468294755566101" }
    ]
  },
  {
    "key": "fc", "name": "分类", "value": [
      { "n": "全部", "v": "" },
      { "n": "新闻", "v": "新闻" }, { "n": "体育", "v": "体育" }, { "n": "综艺", "v": "综艺" },
      { "n": "健康", "v": "健康" }, { "n": "生活", "v": "生活" }, { "n": "科教", "v": "科教" },
      { "n": "经济", "v": "经济" }, { "n": "农业", "v": "农业" }, { "n": "法治", "v": "法治" },
      { "n": "军事", "v": "军事" }, { "n": "少儿", "v": "少儿" }, { "n": "动画", "v": "动画" },
      { "n": "纪实", "v": "纪实" }, { "n": "戏曲", "v": "戏曲" }, { "n": "音乐", "v": "音乐" },
      { "n": "影视", "v": "影视" }
    ]
  },
  {
    "key": "fl", "name": "字母", "value": [
      { "n": "全部", "v": "" },
      "A,B,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V,W,X,Y,Z".split(',').map(l => ({ n: l, v: l }))
    ].flat()
  },
  {
    "key": "year", "name": "年份", "value": [
      { "n": "全部", "v": "" },
      ...Array.from({ length: 23 }, (_, i) => ({ n: (2022 - i).toString(), v: (2022 - i).toString() }))
    ]
  },
  {
    "key": "month", "name": "月份", "value": [
      { "n": "全部", "v": "" },
      ...Array.from({ length: 12 }, (_, i) => ({ n: (i + 1).toString().padStart(2, '0'), v: (i + 1).toString().padStart(2, '0') }))
    ]
  }
];

function init(extend) {
  console.log("央视大全爬虫-简化版（直接使用原始hls_url）");
}

function home() {
  return JSON.stringify({
    class: [{ type_name: "央视大全", type_id: "CCTV" }],
    filters: { "CCTV": filtersConfig }
  });
}

function homeVod() {
  return JSON.stringify({ list: [] });
}

function category(tid, pg, filter, extend) {
  pg = parseInt(pg) || 1;
  let params = { ...(filter || {}), ...(extend || {}) };
  let year = params.year || '';
  let month = params.month || '';
  let prefix = year + month;
  
  let queryParams = {
    fl: params.fl || '',
    fc: params.fc || '',
    cid: params.cid || '',
    p: pg,
    n: 20,
    serviceId: 'tvcctv',
    t: 'json'
  };
  let url = 'https://api.cntv.cn/lanmu/columnSearch?' + Object.entries(queryParams).map(([k, v]) => `${k}=${encodeURIComponent(v)}`).join('&');
  
  let json = fetchSync(url, { json: true });
  if (!json || !json.response || !json.response.docs) {
    return JSON.stringify({ list: [], page: pg, pagecount: 0, total: 0 });
  }
  
  let videos = [];
  for (let vod of json.response.docs) {
    let lastVideo = vod.lastVIDE?.videoSharedCode || '';
    if (lastVideo === '') lastVideo = '_';
    let guid = `${prefix}###${vod.column_name}###${lastVideo}###vod.column_logo || ''}`;
    videos.push({
      vod_id: guid,
      vod_name: vod.column_name,
      vod_pic: vod.column_logo || '',
      vod_remarks: ''
    });
  }
  
  let total = json.response.numFound || videos.length;
  let pagecount = Math.ceil(total / 20);
  return JSON.stringify({
    list: videos,
    page: pg,
    pagecount: pagecount,
    limit: 20,
    total: total
  });
}

// 直接返回原始 hls_url（不再解析子流）
function getRawHlsUrl(pid) {
  if (!pid) return null;
  let infoUrl = `https://vdn.apps.cntv.cn/api/getHttpVideoInfo.do?pid=${pid}`;
  let info = fetchSync(infoUrl, { json: true });
  if (!info || !info.hls_url) return null;
  return info.hls_url.trim();
}

function detail(vodId) {
  let parts = vodId.split('###');
  if (parts.length < 4) return JSON.stringify({ list: [] });
  let prefix = parts[0];
  let title = parts[1];
  let lastVideo = parts[2];
  let logo = parts[3];
  
  if (lastVideo === '_') return JSON.stringify({ list: [] });
  
  let infoUrl = `https://api.cntv.cn/video/videoinfoByGuid?guid=${lastVideo}&serviceId=tvcctv`;
  let infoJson = fetchSync(infoUrl, { json: true });
  if (!infoJson || !infoJson.ctid) return JSON.stringify({ list: [] });
  let topicId = infoJson.ctid;
  let channel = infoJson.channel || '';
  
  let listUrl = `https://api.cntv.cn/NewVideo/getVideoListByColumn?id=${topicId}&d=${prefix}&p=1&n=100&sort=desc&mode=0&serviceId=tvcctv&t=json`;
  let listJson = fetchSync(listUrl, { json: true });
  if (!listJson || !listJson.data || !listJson.data.list) return JSON.stringify({ list: [] });
  
  let videoList = [];
  for (let video of listJson.data.list) {
    let playId = video.pid || video.vid || video.guid;
    if (playId) {
      videoList.push(`${video.title}$${playId}`);
    }
  }
  if (videoList.length === 0) return JSON.stringify({ list: [] });
  
  // 获取第一集的原始 hls_url 用于调试（放入备注字段，避免截断）
  let debugUrl = '';
  let firstPid = videoList[0].split('$')[1];
  if (firstPid) {
    debugUrl = getRawHlsUrl(firstPid) || '';
  }
  
  let displayDate = prefix || new Date().getFullYear().toString();
  let vod = {
    vod_id: vodId,
    vod_name: `${displayDate} ${title}`,
    vod_pic: logo,
    type_name: channel,
    vod_year: displayDate,
    vod_area: "",
    vod_remarks: debugUrl,   // 原始播放地址放入备注，显示更完整
    vod_actor: "",
    vod_director: topicId,   // 恢复为原来的 topicId
    vod_content: "当前页面默认只展示最新100期的内容,可在分类页面选择年份和月份进行往期节目查看。年份和月份仅影响当前页面内容,不参与分类过滤。视频默认播放使用原始hls_url（包含maxbr参数）。",
    vod_play_from: "CCTV",
    vod_play_url: videoList.join("#")
  };
  return JSON.stringify({ list: [vod] });
}

function play(flag, id, vipFlags) {
  let rawUrl = getRawHlsUrl(id);
  if (!rawUrl) {
    return JSON.stringify({ parse: 0, playUrl: '', url: id });
  }
  return JSON.stringify({ parse: 0, playUrl: '', url: rawUrl });
}

function search(wd, quick) {
  return JSON.stringify({ list: [] });
}

__JS_SPIDER__ = { init, home, homeVod, category, detail, play, search };

File: CCTV.py
# -*- coding: utf-8 -*-
import sys
sys.path.append('..')
from base.spider import Spider
import requests
import json
import re
import time
from functools import lru_cache

class cctvSpider(Spider):
    """央视栏目爬虫 - 直接调用央视官方 API"""
    
    def __init__(self):
        self.cache = {}          # 简单缓存
        self.cache_ttl = 600     # 10分钟
    
    def _fetch_json(self, url, timeout=10):
        """请求 JSON 并返回字典"""
        try:
            resp = requests.get(url, timeout=timeout)
            if resp.status_code == 200:
                return resp.json()
        except Exception as e:
            print(f"请求失败: {url} - {e}")
        return {}
    
    def _get_column_list(self, p, n):
        """获取栏目列表（带缓存）"""
        cache_key = f"col_list_{p}_{n}"
        now = time.time()
        if cache_key in self.cache and now - self.cache[cache_key]['ts'] < self.cache_ttl:
            return self.cache[cache_key]['data']
        url = f'https://api.cntv.cn/lanmu/columnSearch?p={p}&n={n}&serviceId=tvcctv&t=json'
        data = self._fetch_json(url)
        docs = data.get('response', {}).get('docs', [])
        result = []
        for item in docs:
            title = item.get('column_name', '无标题')
            last_video = item.get('lastVIDE') or item.get('videoSharedCode')
            if not last_video:
                continue
            pic = item.get('column_logo', '')
            # 将最新视频ID作为栏目标识，缓存栏目信息供 detail 使用
            self.cache[f"col_info_{last_video}"] = {'title': title, 'pic': pic}
            result.append({
                'vod_id': last_video,           # 用最新视频ID作为栏目ID
                'vod_name': title,
                'vod_pic': pic
            })
        self.cache[cache_key] = {'ts': now, 'data': result}
        return result
    
    # ================== 必须实现的接口 ==================
    
    def init(self, ext):
        """初始化（可配置扩展）"""
        print("央视爬虫初始化成功")
        return {}
    
    def home(self, filter=False):
        """返回分类列表（首页分类）"""
        classes = [
            {'type_id': 'cctv', 'type_name': '📺 央视栏目'}
        ]
        # 如果需要筛选器可在此添加
        filters = {}
        return {'class': classes, 'filters': filters}
    
    def homeVod(self):
        """首页推荐（可选）"""
        # 返回空列表，或者可以返回一些热门栏目
        return {'list': []}
    
    def category(self, tid, pg, filter, extend):
        """根据分类ID获取视频列表"""
        # 这里tid固定为 'cctv'，我们直接获取栏目列表
        try:
            p = int(pg) if pg else 1
            n = 20   # 每页数量
            items = self._get_column_list(p, n)
            # 构造返回数据
            total = len(items)   # 这里简单处理，真实API可能返回总数
            return {
                'page': p,
                'pagecount': 1,   # 仅简单分页
                'limit': n,
                'total': total,
                'list': items
            }
        except Exception as e:
            print(f"category 错误: {e}")
            return {'list': []}
    
    def detail(self, tid):
        """获取一个栏目的剧集详情（视频列表）"""
        vid = tid   # tid就是最新视频ID
        # 1. 获取视频信息以得到栏目ctid
        info_url = f'https://api.cntv.cn/video/videoinfoByGuid?guid={vid}&serviceId=tvcctv'
        info = self._fetch_json(info_url)
        ctid = info.get('ctid')
        channel = info.get('channel', '')
        if not ctid:
            return {'list': []}
        # 2. 获取该栏目下的视频列表
        list_url = f'https://api.cntv.cn/NewVideo/getVideoListByColumn?id={ctid}&d=&p=1&n=100&sort=desc&mode=0&serviceId=tvcctv&t=json'
        list_data = self._fetch_json(list_url)
        video_list = list_data.get('data', {}).get('list', [])
        # 构建播放链接字符串
        play_urls = []
        for video in video_list:
            play_id = video.get('pid') or video.get('vid') or video.get('guid')
            if play_id:
                title = video.get('title', '未命名')
                play_urls.append(f"{title}${play_id}")
        if not play_urls:
            return {'list': []}
        # 从缓存获取栏目名和图片（之前在列表时已缓存）
        col_info = self.cache.get(f"col_info_{vid}", {})
        title = col_info.get('title', '央视栏目')
        pic = col_info.get('pic', '')
        vod = {
            'vod_id': tid,
            'vod_name': title,
            'vod_pic': pic,
            'type_name': channel,
            'vod_play_from': 'CCTV',
            'vod_play_url': '#'.join(play_urls)
        }
        return {'list': [vod]}
    
    def play(self, flag, id, vipFlags):
        """获取播放地址"""
        # id 就是视频的 pid/vid
        play_url = f'https://vdn.apps.cntv.cn/api/getHttpVideoInfo.do?pid={id}'
        data = self._fetch_json(play_url)
        hls = data.get('hls_url', '')
        return {'parse': 0, 'playUrl': '', 'url': hls}
    
    def search(self, wd, quick):
        """搜索功能（可选）"""
        # 这里可以调用央视搜索接口，但暂时返回空
        return {'list': []}

# 导出类名（TVBox 要求）
Spider = cctvSpider